File: day9/rust/src/main.rs
use itertools::Itertools;

fn block_hash(start_pos: u32, size: u32, id: u32) -> u64 {
    let start_pos = start_pos as u64;
    let size = size as u64;
    let id = id as u64;

    if start_pos > 0 {
        id * ((start_pos + size - 1) * (start_pos + size) - (start_pos - 1) * start_pos) / 2
    } else {
        id * ((start_pos + size - 1) * (start_pos + size)) / 2
    }
}

fn part1(disk_map: &Vec<u32>) -> u64 {
    let mut disk_map = disk_map.clone();
    let mut current_pos = 0;
    let mut current_idx = 0;
    let mut hash = 0;
    loop {
        if current_idx >= disk_map.len() {
            break;
        }
        let size = disk_map[current_idx];

        if current_idx % 2 == 0 {
            let id = current_idx as u32 / 2;
            hash += block_hash(current_pos, size, id);
            current_pos += size;
        } else {
            let mut gap = size;
            while gap > 0 {
                let last = disk_map.last().unwrap();
                let last_id = (disk_map.len() - 1) as u32 / 2;
                if gap >= *last {
                    gap -= last;
                    hash += block_hash(current_pos, *last, last_id);
                    current_pos += last;
                    disk_map.pop(); // pop item
                    disk_map.pop(); // pop gap
                } else {
                    hash += block_hash(current_pos, gap, last_id);
                    current_pos += gap;
                    *disk_map.last_mut().unwrap() -= gap;
                    break;
                }
            }
        }
        current_idx += 1;
    }
    hash
}

fn part2(disk_map: &Vec<u32>) -> u64 {
    let disk_map = disk_map.clone();
    let mut hash = 0;
    let mut indices: Vec<u32> = vec![0];
    let mut gaps: Vec<(u32, u32)> = Vec::new();

    for (id, (block, gap)) in disk_map.iter().tuples().enumerate() {
        indices.push(indices[id] + block + gap);
        let prev_gap = gaps.last().unwrap_or(&(0, 0));
        gaps.push((prev_gap.0 + prev_gap.1 + block, *gap));
    }

    for cand_id in (0..((disk_map.len() + 1) / 2)).rev() {
        let cand_map_idx = cand_id * 2;
        let cand_size = disk_map[cand_map_idx];
        let mut new_cand_idx = indices[cand_id];
        for (gap_idx, gap) in gaps.iter_mut().enumerate() {
            if indices[cand_id] < gap.0 {
                // no gaps to the left
                break;
            }
            if cand_size <= gap.1 {
                new_cand_idx = gap.0;
                gap.1 -= cand_size;
                gap.0 += cand_size;
                if gap.1 == 0 {
                    gaps.remove(gap_idx); // theoretically this should be O(1), but rust doesn't have a suitable linked list
                }
                break;
            }
        }
        hash += block_hash(new_cand_idx, cand_size, cand_id as u32);
    }
    hash
}

fn main() {
    let disk_map: Vec<u32> = std::fs::read_to_string("../diskmap.txt")
        .unwrap()
        .chars()
        .map(|c| c.to_digit(10).unwrap())
        .collect();
    println!("Part 1: {}", part1(&disk_map));
    println!("Part 2: {}", part2(&disk_map));
}

File: day9/haskell/Main.hs
import Data.Char (digitToInt)
import Data.Int (Int64)

blockHash :: Int64 -> Int64 -> Int64 -> Int64
blockHash 0 size id = id * quot ((size - 1) * size) 2
blockHash start size id = id * quot ((start + size - 1) * (start + size) - (start - 1) * start) 2

processBlock :: [Int64] -> Int64 -> Int64 -> Int64 -> Int64 -> Int64
processBlock [] current_pos left_id right_id hash = hash
processBlock [block] current_pos left_id right_id hash = hash + blockHash current_pos block left_id
processBlock (block:xs) current_pos left_id right_id hash = processGap xs (current_pos + block) (left_id + 1) right_id (hash + blockHash current_pos block left_id)

processGap :: [Int64] -> Int64 -> Int64 -> Int64 -> Int64 -> Int64
processGap [] current_pos left_id right_id hash = hash
processGap [gap, block] current_pos left_id right_id hash = hash + blockHash current_pos gap right_id
processGap disk current_pos left_id right_id hash
    | block > gap  = processBlock (init notfirst ++ [block - gap])     (current_pos + gap)   left_id right_id       (hash + blockHash current_pos gap right_id)
    | block == gap = processBlock ((init.init) notfirst)               (current_pos + gap)   left_id (right_id - 1) (hash + blockHash current_pos gap right_id)
    | block < gap  = processGap   ((gap - block):(init.init) notfirst) (current_pos + block) left_id (right_id - 1) (hash + blockHash current_pos block right_id)
    where
        (gap:notfirst) = disk
        block = last disk
    
part1 :: [Int64] -> Int64
part1 disk = processBlock disk 0 0 (((fromIntegral.length) disk - 1) `quot` 2) 0

extractGaps :: [Int64] -> [(Int64, Int64)]

main::IO()
main = do
  disk_map_str <- readFile "../diskmap.txt"
  -- let disk_map_str = "2333133121414131402"
  let disk_map :: [Int64] = map (fromIntegral.digitToInt) disk_map_str
  print (part1 disk_map)
